internal/controller/certificates/policies/checks.go

```diff
@@ -94,6 +94,33 @@ func SecretPrivateKeyMatchesSpec(input Input) (string, string, bool) {
 	return "", "", false
 }
 
+func SecretKeystoreFormatMatchesSpec(input Input) (string, string, bool) {
+	if input.Certificate.Spec.Keystores == nil {
+		if len(input.Secret.Data[cmapi.Pkcs12SecretKey]) == 0 ||
+			len(input.Secret.Data[cmapi.Pkcs12TruststoreKey]) == 0 ||
+			len(input.Secret.Data[cmapi.JksSecretKey]) == 0 ||
+			len(input.Secret.Data[cmapi.JksTruststoreKey]) == 0 {
+			return SecretMismatch, "Keystore keys does not contain data", true
+		}
+		return "", "", false
+	}
+
+	if input.Certificate.Spec.Keystores.JKS != nil {
+		if len(input.Secret.Data[cmapi.JksSecretKey]) == 0 ||
+			len(input.Secret.Data[cmapi.JksTruststoreKey]) == 0 {
+			return SecretMismatch, "JKS Keystore keys does not contain data", true
+		}
+	}
+
+	if input.Certificate.Spec.Keystores.PKCS12 != nil {
+		if len(input.Secret.Data[cmapi.Pkcs12SecretKey]) == 0 ||
+			len(input.Secret.Data[cmapi.Pkcs12TruststoreKey]) == 0 {
+			return SecretMismatch, "PKCS12 Keystore keys does not contain data", true
+		}
+	}
+	return "", "", false
+}
+
```

Comment:
What happens if private key rotation is set to `Never` and the `KeyStores` setting changes? Does cert-manager read the existing private key (in which ever format) and re-encode it it in the new desired keystore format.
 * https://cert-manager.io/docs/usage/certificate/#rotation-private-key

If you can find our what the behaviour is, please add a few comments to the code explaining it for the benefit of future maintainers.

Reply:
It is the later, it re-encodes the new desired keystore format! I will add some comments in the code.